test/e2e/webdriver/driver.js

```diff
@@ -862,6 +862,47 @@ class Driver {
     return await this.windowHandles.getAllWindowHandles();
   }
 
+  /**
+   * Function that aims to simulate a click action on a specified web element
+   * within a web page and waits for the current window to close.
+   *
+   * @param {string | object} rawLocator - Element locator
+   * @param {number} [retries] - The number of times to retry the click action if it fails
+   * @returns {Promise<void>} promise that resolves to the WebElement
+   */
+  async clickElementAndWaitForWindowToClose(rawLocator, retries = 3) {
+    const handle = await this.driver.getWindowHandle();
+    await this.clickElement(rawLocator, retries);
+    await this.waitForWindowToClose(handle);
+  }
+
+  /**
+   * Waits for the specified window handle to close before returning.
+   *
+   * @param {string} handle - The handle of the window or tab we'll wait for.
+   * @param {number} [timeout] - The amount of time in milliseconds to wait
+   * before timing out. Defaults to `this.timeout`.
+   * @throws {Error} throws an error if the window handle doesn't close within
+   * the timeout.
+   */
+  async waitForWindowToClose(handle, timeout = this.timeout) {
+    const start = Date.now();
+    // eslint-disable-next-line no-constant-condition
+    while (true) {
+      const handles = await this.getAllWindowHandles();
+      if (!handles.includes(handle)) {
+        return;
+      }
+
+      const timeElapsed = Date.now() - start;
+      if (timeElapsed > timeout) {
+        throw new Error(
+          `waitForWindowToClose timed out waiting for window handle '${handle}' to close.`,
+        );
+      }
+    }
+  }
```

Comment:
How do you feel about adding a title?  It would allow for slightly easier-to-read error messages.

```suggestion
  async clickElementAndWaitForWindowToClose(rawLocator, retries = 3) {
    const handle = await this.driver.getWindowHandle();
    const title = await this.driver.getTitle();
    await this.clickElement(rawLocator, retries);
    await this.waitForWindowToClose(handle, title);
  }

  /**
   * Waits for the specified window handle to close before returning.
   *
   * @param {string} handle - The handle of the window or tab we'll wait for.
   * @param {number} [timeout] - The amount of time in milliseconds to wait
   * before timing out. Defaults to `this.timeout`.
   * @throws {Error} throws an error if the window handle doesn't close within
   * the timeout.
   */
  async waitForWindowToClose(handle, title, timeout = this.timeout) {
    const start = Date.now();
    // eslint-disable-next-line no-constant-condition
    while (true) {
      const handles = await this.getAllWindowHandles();
      if (!handles.includes(handle)) {
        return;
      }

      const timeElapsed = Date.now() - start;
      if (timeElapsed > timeout) {
        throw new Error(
          `waitForWindowToClose timed out waiting for window handle '${handle}' with title '${title}' to close.`,
        );
      }
    }
  }
```

Reply:
I went down this route to add the title to the error message, but I found it likely to be flaky. The way it's prototyped above, where `title` isn't guaranteed to be the actual title of the window due to race conditions between multiple async calls to query the windows (this async nature of these calls seem to be the cause of another race condition in our tests), seems especially prone to flakiness. 

That said, _all_ of these async window interactions are potentially flaky since the windows can be changing while we are interacting with them. I'm in favor of removing as much potential flakiness as possible from the hot path.

If the async query for the current window title were to be made in the error path of `waitForWindowToClose` it might be slightly better... but a title fetched by `getTitle` isn't _guaranteed_ to match that of the referenced window handle (since the `windowHandle` could finally close during the call to `getTitle`, in which case it would return a different title).

All this just makes me want to switch to a better browser testing framework, if there is one, and/or extend `socket-background.to-mocha.ts` to automatically sync all tab state (`chrome.tabs.query`) with the server, and implement event-driven interactions (e.g., `windowHandle.on("close", ...)`) and maybe even some sort of interaction `lock` mechanism.